cmd/keytransparency-client/cmd/hammer.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/google/keytransparency/core/authentication"
+
+	"github.com/aybabtme/uniplot/histogram"
+	"github.com/paulbellamy/ratecounter"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+
+	tm "github.com/buger/goterm"
+	tpb "github.com/google/keytransparency/core/api/type/type_proto"
+)
+
+var (
+	maxWorkers int
+	workers    int
+	memLog     = new(bytes.Buffer)
+	ramp       time.Duration
+)
+
+func init() {
+	log.SetOutput(memLog)
+
+	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
+
+	RootCmd.AddCommand(hammerCmd)
+
+	hammerCmd.Flags().IntVar(&maxWorkers, "workers", 1, "Number of parallel workers")
+	hammerCmd.Flags().DurationVar(&ramp, "ramp", 1*time.Second, "Time to spend ramping up")
+}
+
+// hammerCmd represents the post command
+var hammerCmd = &cobra.Command{
+	Use:   "hammer",
+	Short: "Loadtest the server",
+	Long:  ``,
+
+	PreRun: func(cmd *cobra.Command, args []string) {
+		if err := readKeyStoreFile(); err != nil {
+			log.Fatal(err)
+		}
+	},
+	RunE: func(cmd *cobra.Command, args []string) error {
+		ctx := context.Background()
+		parallel(ctx, maxWorkers)
+		return nil
+	},
+}
+
+func parallel(ctx context.Context, maxWorkers int) {
+	times := make(chan time.Duration)
+	jobs := make(chan job)
+	var wg sync.WaitGroup
+	go recordLatencies(ctx, times)
+	go generateJobs(ctx, jobs)
+
+	// Slowly add workers up to maxWorkers
+	ramp := time.NewTicker(time.Duration(ramp.Nanoseconds() / int64(maxWorkers)))
+	for ; workers < maxWorkers && ctx.Err() == nil; <-ramp.C {
```

Comment:
I think this might break if someone sets maxWorkers flag negative.

Reply:
I'll make it uint